traffic_ops/traffic_ops_golang/servicecategory/servicecategories.go

```diff
@@ -0,0 +1,199 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"fmt"
+	"github.com/jmoiron/sqlx"
+	"net/http"
+	"strconv"
+
+	"github.com/apache/trafficcontrol/lib/go-log"
+	"github.com/apache/trafficcontrol/lib/go-tc"
+	"github.com/apache/trafficcontrol/lib/go-tc/tovalidate"
+	"github.com/apache/trafficcontrol/lib/go-util"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/auth"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/dbhelpers"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/tenant"
+
+	validation "github.com/go-ozzo/ozzo-validation"
+)
+
+//we need a type alias to define functions on
+type TOServiceCategory struct {
+	api.APIInfoImpl `json:"-"`
+	tc.ServiceCategoryNullable
+}
+
+func (v *TOServiceCategory) SetLastUpdated(t tc.TimeNoMod) { v.LastUpdated = &t }
+func (v *TOServiceCategory) InsertQuery() string           { return insertQuery() }
+func (v *TOServiceCategory) NewReadObj() interface{}       { return &tc.ServiceCategory{} }
+func (v *TOServiceCategory) SelectQuery() string           { return selectQuery() }
+func (v *TOServiceCategory) UpdateQuery() string           { return updateQuery() }
+func (v *TOServiceCategory) DeleteQuery() string           { return deleteQuery() }
+
+func (serviceCategory TOServiceCategory) GetAuditName() string {
+	if serviceCategory.Name != nil {
+		return *serviceCategory.Name
+	}
+	if serviceCategory.ID != nil {
+		return strconv.Itoa(*serviceCategory.ID)
+	}
+	if serviceCategory.TenantID != nil {
+		return strconv.Itoa(*serviceCategory.TenantID)
+	}
+	return "unknown"
+}
+
+func (serviceCategory TOServiceCategory) GetKeyFieldsInfo() []api.KeyFieldInfo {
+	return []api.KeyFieldInfo{{"id", api.GetIntKey}}
+}
+
+//Implementation of the Identifier, Validator interface functions
+func (serviceCategory TOServiceCategory) GetKeys() (map[string]interface{}, bool) {
+	if serviceCategory.ID == nil {
+		return map[string]interface{}{"id": 0}, false
+	}
+	return map[string]interface{}{"id": *serviceCategory.ID}, true
+}
+
+func (serviceCategory *TOServiceCategory) SetKeys(keys map[string]interface{}) {
+	i, _ := keys["id"].(int) //this utilizes the non panicking type assertion, if the thrown away ok variable is false i will be the zero of the type, 0 here.
+	serviceCategory.ID = &i
+}
+
+func (serviceCategory TOServiceCategory) GetType() string {
+	return "serviceCategory"
+}
+
+func (serviceCategory TOServiceCategory) Validate() error {
+	errs := validation.Errors{
+		"name":     validation.Validate(serviceCategory.Name, validation.NotNil, validation.Required),
+		"tenantId": validation.Validate(serviceCategory.TenantID, validation.Min(1)),
```

Comment:
This isn't everything that needs to be validated here. Specifically, that tenant has to exist.

btw, I've been mostly holding off on mentioning things like this until it's fixed in the database, but there are some other places where this kind of thing needs to be checked.